source/grid/grid_tools_topology.cc

```diff
@@ -1855,8 +1855,9 @@ namespace GridTools
     const Triangulation<dim, spacedim> &triangulation,
     DynamicSparsityPattern             &cell_connectivity)
   {
-    std::vector<std::vector<unsigned int>> vertex_to_cell(
-      triangulation.n_vertices());
+    // Assume at most 16 neighbors per cell (a structured hex mesh has 8)
+    std::vector<boost::container::small_vector<unsigned int, 16>>
+      vertex_to_cell(triangulation.n_vertices());
```

Comment:
Out of curiosity, how many neighbors does a vertex have on a structured *tet* mesh?

Reply:
I'm curious about this too. I think that, for structured meshes, it depends significantly on how you generate them but I am more interested in unstructured meshes (if we can use structured grids we can probably also use hexahedra). I computed the valence for all vertices and internal vertices of one of my heart meshes:

```cpp
#include <deal.II/grid/grid_in.h>
#include <deal.II/grid/tria.h>

#include <iostream>
#include <map>
#include <set>

int
main()
{
  using namespace dealii;

  Triangulation<3> tria;
  GridIn<3>        grid_in(tria);
  grid_in.read_exodusii("heart.e");
  std::cout << "Number of cells     = " << tria.n_active_cells() << std::endl;
  std::cout << "Number of vertices  = " << tria.n_vertices() << std::endl;

  std::map<unsigned int, std::set<unsigned int>> vertex_neighbors;
  std::map<unsigned int, std::set<unsigned int>> internal_vertex_neighbors;
  for (const auto &cell : tria.active_cell_iterators())
    {
      for (const auto line_no : cell->line_indices())
        {
          vertex_neighbors[cell->line(line_no)->vertex_index(0)].insert(
            cell->line(line_no)->vertex_index(1));
          vertex_neighbors[cell->line(line_no)->vertex_index(1)].insert(
            cell->line(line_no)->vertex_index(0));
        }
      if (!cell->at_boundary())
        for (const auto line_no : cell->line_indices())
          {
            internal_vertex_neighbors[cell->line(line_no)->vertex_index(0)]
              .insert(cell->line(line_no)->vertex_index(1));
            internal_vertex_neighbors[cell->line(line_no)->vertex_index(1)]
              .insert(cell->line(line_no)->vertex_index(0));
          }
    }

  unsigned int mean_n_neighbors          = 0;
  unsigned int internal_mean_n_neighbors = 0;
  for (const auto &pair : vertex_neighbors)
    mean_n_neighbors += pair.second.size();
  for (const auto &pair : internal_vertex_neighbors)
    internal_mean_n_neighbors += pair.second.size();

  std::cout << "mean vertex valence = "
            << mean_n_neighbors / vertex_neighbors.size() << std::endl;
  std::cout << "internal mean vertex valence = "
            << internal_mean_n_neighbors / internal_vertex_neighbors.size()
            << std::endl;
}
```

computes an average number of 12 neighbors per vertex and 11 neighbors per internal vertex.

Reply:
Ah, interesting experiment. From a conceptual perspective, you probably want to set the size of the small vector so that it covers perhaps 95% of vertices. Do you happen to have the distribution of the number of neighbors per vertex?

Reply:
I added the following bit of code:
```cpp
  std::map<unsigned int, unsigned int> neighbor_counts;
  for (const auto &pair : vertex_neighbors)
    {
      mean_n_neighbors += pair.second.size();
      neighbor_counts[pair.second.size()] += 1;
    }
```
and I see the distribution
```
3, 1
4, 728
5, 4084
6, 7614
7, 17329
8, 31145
9, 46698
10, 64193
11, 68269
12, 63574
13, 57016
14, 50476
15, 41886
16, 31820
17, 21269
18, 12217
19, 6072
20, 2527
21, 825
22, 262
23, 61
24, 12
26, 1
```
so, roughly speaking, 16 looks like a decent guess for tets too if we want to cover the majority of cases.

Reply:
I would accept a patch that would add the table above as a comment to document why the choice of 16 is reasonable.